[Bug]: Error: The terraform-provider-aws_v5.65.0_x5 plugin crashed!

### Terraform Core Version

1.9.5

### AWS Provider Version

5.65

### Affected Resource(s)

aws_vpc_dhcp_options

### Expected Behavior

the plan is getting successfull, so ideally the resource should have been created but insted it threw me error

### Actual Behavior

Plugin crashed on apply while creating the resource aws vpc dhcp options.

### Relevant Error/Panic Output Snippet

```shell
Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc_dhcp_options.dhcp_options["config1"],
│   on terraform-aws-vpc/main.tf line 13, in resource "aws_vpc_dhcp_options" "dhcp_options":
│   13: resource "aws_vpc_dhcp_options" "dhcp_options" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc.vpc,
│   on terraform-aws-vpc/main.tf line 1, in resource "aws_vpc" "vpc":
│    1: resource "aws_vpc" "vpc" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-aws_v5.65.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 90 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.(*dhcpOptionsMap).resourceDataToDHCPConfigurations(0xc003135110?, 0xc002001580)
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:301 +0x657
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.resourceVPCDHCPOptionsCreate({0x17dffe88, 0xc0028348a0}, 0xc002001580, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:113 +0xa5
github.com/hashicorp/terraform-provider-aws/internal/provider.New.(*wrappedResource).Create.interceptedHandler[...].func8(0xc002001580?, {0x15d3bce0?, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:113 +0x283
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x17dffe88?, {0x17dffe88?, 0xc00281d4d0?}, 0xd?, {0x15d3bce0?, 0xc003135110?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:801 +0x7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00102e2a0, {0x17dffe88, 0xc00281d4d0}, 0xc0028265b0, 0xc002001400, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0xa69
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc004317140, {0x17dffe88?, 0xc00281d410?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc000dd17a0, {0x17dffe88?, 0xc00281d140?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001d40be0, {0x17dffe88?, 0xc00281c780?}, 0xc001b61730)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x3bc
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x15b56640, 0xc001d40be0}, {0x17dffe88, 0xc00281c780}, 0xc002000c00, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003c5600, {0x17dffe88, 0xc00281c6f0}, {0x17e6e0a0, 0xc003244000}, 0xc002824b40, 0xc0031df4a0, 0x217cb5b8, 0x0)
        google.golang.org/grpc@v1.66.0/server.go:1393 +0xe11
google.golang.org/grpc.(*Server).handleStream(0xc0003c5600, {0x17e6e0a0, 0xc003244000}, 0xc002824b40)
        google.golang.org/grpc@v1.66.0/server.go:1804 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.0/server.go:1029 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        google.golang.org/grpc@v1.66.0/server.go:1040 +0x125


Error: The terraform-provider-aws_v5.65.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform apply (exit 1)
```


### Terraform Configuration Files

```hcl
 terraform {
   required_providers {
     aws = {
       source  = "hashicorp/aws"
       version = "~> 5"
     }
   }
 
   cloud {
     organization = "Planview"
     workspaces {
       name = "<workspace_name>"
     }
   }
 }
```

### Steps to Reproduce

terraform init
terraform plan
terraform apply

### Debug Output

```
Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc_dhcp_options.dhcp_options["config1"],
│   on terraform-aws-vpc/main.tf line 13, in resource "aws_vpc_dhcp_options" "dhcp_options":
│   13: resource "aws_vpc_dhcp_options" "dhcp_options" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc.vpc,
│   on terraform-aws-vpc/main.tf line 1, in resource "aws_vpc" "vpc":
│    1: resource "aws_vpc" "vpc" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-aws_v5.65.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 90 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.(*dhcpOptionsMap).resourceDataToDHCPConfigurations(0xc003135110?, 0xc002001580)
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:301 +0x657
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.resourceVPCDHCPOptionsCreate({0x17dffe88, 0xc0028348a0}, 0xc002001580, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:113 +0xa5
github.com/hashicorp/terraform-provider-aws/internal/provider.New.(*wrappedResource).Create.interceptedHandler[...].func8(0xc002001580?, {0x15d3bce0?, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:113 +0x283
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x17dffe88?, {0x17dffe88?, 0xc00281d4d0?}, 0xd?, {0x15d3bce0?, 0xc003135110?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:801 +0x7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00102e2a0, {0x17dffe88, 0xc00281d4d0}, 0xc0028265b0, 0xc002001400, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0xa69
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc004317140, {0x17dffe88?, 0xc00281d410?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc000dd17a0, {0x17dffe88?, 0xc00281d140?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001d40be0, {0x17dffe88?, 0xc00281c780?}, 0xc001b61730)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x3bc
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x15b56640, 0xc001d40be0}, {0x17dffe88, 0xc00281c780}, 0xc002000c00, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003c5600, {0x17dffe88, 0xc00281c6f0}, {0x17e6e0a0, 0xc003244000}, 0xc002824b40, 0xc0031df4a0, 0x217cb5b8, 0x0)
        google.golang.org/grpc@v1.66.0/server.go:1393 +0xe11
google.golang.org/grpc.(*Server).handleStream(0xc0003c5600, {0x17e6e0a0, 0xc003244000}, 0xc002824b40)
        google.golang.org/grpc@v1.66.0/server.go:1804 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.0/server.go:1029 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        google.golang.org/grpc@v1.66.0/server.go:1040 +0x125


Error: The terraform-provider-aws_v5.65.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform apply (exit 1)

### Panic Output

Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc_dhcp_options.dhcp_options["config1"],
│   on terraform-aws-vpc/main.tf line 13, in resource "aws_vpc_dhcp_options" "dhcp_options":
│   13: resource "aws_vpc_dhcp_options" "dhcp_options" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
╷
│ Error: Plugin did not respond
│
│   with module.vpc["vpc_1"].aws_vpc.vpc,
│   on terraform-aws-vpc/main.tf line 1, in resource "aws_vpc" "vpc":
│    1: resource "aws_vpc" "vpc" {
│
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵

Stack trace from the terraform-provider-aws_v5.65.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 90 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.(*dhcpOptionsMap).resourceDataToDHCPConfigurations(0xc003135110?, 0xc002001580)
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:301 +0x657
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.resourceVPCDHCPOptionsCreate({0x17dffe88, 0xc0028348a0}, 0xc002001580, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/service/ec2/vpc_dhcp_options.go:113 +0xa5
github.com/hashicorp/terraform-provider-aws/internal/provider.New.(*wrappedResource).Create.interceptedHandler[...].func8(0xc002001580?, {0x15d3bce0?, 0xc003135110})
        github.com/hashicorp/terraform-provider-aws/internal/provider/intercept.go:113 +0x283
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x17dffe88?, {0x17dffe88?, 0xc00281d4d0?}, 0xd?, {0x15d3bce0?, 0xc003135110?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:801 +0x7a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00102e2a0, {0x17dffe88, 0xc00281d4d0}, 0xc0028265b0, 0xc002001400, {0x15d3bce0, 0xc003135110})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/resource.go:937 +0xa69
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc004317140, {0x17dffe88?, 0xc00281d410?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/schema/grpc_provider.go:1153 +0xd5c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ApplyResourceChange(0xc000dd17a0, {0x17dffe88?, 0xc00281d140?}, 0xc001e8c500)
        github.com/hashicorp/terraform-plugin-mux@v0.16.0/tf5muxserver/mux_server_ApplyResourceChange.go:36 +0x193
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001d40be0, {0x17dffe88?, 0xc00281c780?}, 0xc001b61730)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/tf5server/server.go:865 +0x3bc
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x15b56640, 0xc001d40be0}, {0x17dffe88, 0xc00281c780}, 0xc002000c00, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:518 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003c5600, {0x17dffe88, 0xc00281c6f0}, {0x17e6e0a0, 0xc003244000}, 0xc002824b40, 0xc0031df4a0, 0x217cb5b8, 0x0)
        google.golang.org/grpc@v1.66.0/server.go:1393 +0xe11
google.golang.org/grpc.(*Server).handleStream(0xc0003c5600, {0x17e6e0a0, 0xc003244000}, 0xc002824b40)
        google.golang.org/grpc@v1.66.0/server.go:1804 +0xe8b
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.66.0/server.go:1029 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        google.golang.org/grpc@v1.66.0/server.go:1040 +0x125


Error: The terraform-provider-aws_v5.65.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Operation failed: failed running terraform apply (exit 1)
```

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

--- a/.changelog/39427.txt
+++ b/.changelog/39427.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_vpc_dhcp_options: Fix a bug causing a panic crash when an option is absent
+```

--- a/internal/service/ec2/vpc_dhcp_options.go
+++ b/internal/service/ec2/vpc_dhcp_options.go
@@ -255,27 +255,33 @@ func newDHCPOptionsMap(tfToApi map[string]string) *dhcpOptionsMap {
 
 // dhcpConfigurationsToResourceData sets Terraform ResourceData from a list of AWS API DHCP configurations.
 func (m *dhcpOptionsMap) dhcpConfigurationsToResourceData(dhcpConfigurations []awstypes.DhcpConfiguration, d *schema.ResourceData) error {
-	for v := range m.tfToApi {
-		d.Set(v, nil)
+	// Clear existing values
+	for tfName := range m.tfToApi {
+		d.Set(tfName, nil)
 	}
 
 	for _, dhcpConfiguration := range dhcpConfigurations {
 		apiName := aws.ToString(dhcpConfiguration.Key)
-		if tfName, ok := m.apiToTf[apiName]; ok {
-			switch v := d.Get(tfName).(type) {
-			case string:
-				d.Set(tfName, dhcpConfiguration.Values[0].Value)
-			case []interface{}:
-				var values []*string
-				for _, v := range dhcpConfiguration.Values {
-					values = append(values, v.Value)
+		tfName, ok := m.apiToTf[apiName]
+		if !ok {
+			return fmt.Errorf("unsupported DHCP option: %s", apiName)
+		}
+
+		currentValue := d.Get(tfName)
+
+		switch currentValue.(type) {
+		case string:
+			d.Set(tfName, dhcpConfiguration.Values[0].Value)
+		case []interface{}:
+			var values []string
+			for _, v := range dhcpConfiguration.Values {
+				if v.Value != nil {
+					values = append(values, aws.ToString(v.Value))
 				}
-				d.Set(tfName, values)
-			default:
-				return fmt.Errorf("Attribute (%s) is of unsupported type: %T", tfName, v)
 			}
-		} else {
-			return fmt.Errorf("Unsupported DHCP option: %s", apiName)
+			d.Set(tfName, values)
+		default:
+			return fmt.Errorf("attribute (%s) is of unsupported type: %T", tfName, currentValue)
 		}
 	}
 
@@ -287,7 +293,8 @@ func (m *dhcpOptionsMap) resourceDataToDHCPConfigurations(d *schema.ResourceData
 	var output []awstypes.NewDhcpConfiguration
 
 	for tfName, apiName := range m.tfToApi {
-		switch v := d.Get(tfName).(type) {
+		value := d.Get(tfName)
+		switch v := value.(type) {
 		case string:
 			if v != "" {
 				output = append(output, awstypes.NewDhcpConfiguration{
@@ -297,10 +304,9 @@ func (m *dhcpOptionsMap) resourceDataToDHCPConfigurations(d *schema.ResourceData
 			}
 		case []interface{}:
 			var values []string
-			for _, v := range v {
-				v := v.(string)
-				if v != "" {
-					values = append(values, v)
+			for _, item := range v {
+				if str, ok := item.(string); ok && str != "" {
+					values = append(values, str)
 				}
 			}
 			if len(values) > 0 {
@@ -310,7 +316,7 @@ func (m *dhcpOptionsMap) resourceDataToDHCPConfigurations(d *schema.ResourceData
 				})
 			}
 		default:
-			return nil, fmt.Errorf("Attribute (%s) is of unsupported type: %T", tfName, v)
+			return nil, fmt.Errorf("attribute (%s) is of unsupported type: %T", tfName, value)
 		}
 	}
 

